### PR TITLE
Add mutex to ensure only single thread at pybind->c++ layer

### PR DESCRIPTION
**Pull request by phoebusm**
#### Reference Issues/PRs

https://github.com/man-group/ArcticDB/issues/974
#### What does this implement or fix?
https://github.com/man-group/ArcticDB/pull/973#discussion_r1363967704
#### Checklist

<details>
  <summary>
   Checklist for code changes...
  </summary>
 
 - [x] Have you updated the relevant docstrings, documentation and copyright notice?
 - [x] Is this contribution tested against [all ArcticDB's features](../docs/mkdocs/docs/technical/contributing.md)?
 - [x] Do all exceptions introduced raise appropriate [error messages](https://docs.arcticdb.io/error_messages/)?
 - [x] Are API changes highlighted in the PR description?
 - [x] Is the PR labelled as enhancement or bug so it appears in autogenerated release notes?
</details>

### CODE REVIEW

File: cpp/arcticdb/version/version_store_api.cpp
Comment (vasil-pashov) [inline cpp/arcticdb/version/version_store_api.cpp:1129]: Why is there a gil release call here? IMO this does a bit more than the name implies.

**Comment by poodlewars** [comment]
Please could you,

- Explain why we need the GIL part? I assume this is something about what happens if we hold the GIL in the thread blocked by the new mutex, which might deadlock the thread holding the mutex if _that_ has previously dropped the GIL and is now waiting to re-acquire it.
- Confirm how this works with Python multiprocessing? Is there a new mutex for each sub-process?
- Add tests for this for both multiprocessing (should permit the concurrency) and multithreading (should forbid it)

**Comment by phoebusm** [comment]
> Please could you,
> 
> * Explain why we need the GIL part? I assume this is something about what happens if we hold the GIL in the thread blocked by the new mutex, which might deadlock the thread holding the mutex if _that_ has previously dropped the GIL and is now waiting to re-acquire it.
> * Confirm how this works with Python multiprocessing? Is there a new mutex for each sub-process?
> * Add tests for this for both multiprocessing (should permit the concurrency) and multithreading (should forbid it)
> * Share any performance benchmarking showing the impact of this when not using Python multi-threading
> 
> I thought the plan was to add this to all the functions. Is the idea that that will be a follow up?

1. Exactly
2. Correct me if I am wrong, but I believe C++ mutex is intraprocess only. 
3. Yea it passes both multithreading and multiprocessing test
4. Yes this is something I should do

I think I have misunderstood the urgency of the fix. I thought it is an urgent task so I only fix the one being reported. That's why this PR is drafted.
Pending task for a complete PR:
1. Mutex for all pybind function
2. Test
3. Benchmark
If above tasks are required before merging, I will like to add WIP to this PR or simply close it until the above tasks are ready

**Comment by poodlewars** [comment]
> > Please could you,
> > 
> > * Explain why we need the GIL part? I assume this is something about what happens if we hold the GIL in the thread blocked by the new mutex, which might deadlock the thread holding the mutex if _that_ has previously dropped the GIL and is now waiting to re-acquire it.
> > * Confirm how this works with Python multiprocessing? Is there a new mutex for each sub-process?
> > * Add tests for this for both multiprocessing (should permit the concurrency) and multithreading (should forbid it)
> > * Share any performance benchmarking showing the impact of this when not using Python multi-threading
> > 
> > I thought the plan was to add this to all the functions. Is the idea that that will be a follow up?
> 
>     1. Exactly
> 
>     2. Correct me if I am wrong, but I believe C++ mutex is intraprocess only. <- Ok just realise something. Let me do more tests
> 
>     3. Yea it passes both multithreading and multiprocessing test
> 
>     4. Yes this is something I should do
> 
> 
> I think I have misunderstood the urgency of the fix. I thought it is an urgent task so I only fix the functions being reported in this PR. That's why this PR is drafted. Pending task for a complete PR:
> 
>     1. Mutex for all pybind function
> 
>     2. Test
> 
>     3. Benchmark
> 
> 
> If above tasks are required before merging, I will like to add WIP to this PR or simply close it until the above tasks are ready

Thanks. It would be good to confirm the urgency with the users, I'm not sure whether they have a good workaround or not. If they're in a rush, I think it's OK just to add this to read and read batch, and we can follow up immediately for the other methods. Otherwise, let's do them all at once and save ourselves an extra release.

I think it would be good to have the multithreading and processing tests whether they are in a rush or not, we should be careful that we are not making things worse.

File: cpp/arcticdb/version/version_store_api.cpp
Comment (vasil-pashov) [reply cpp/arcticdb/version/version_store_api.cpp:1129]: Can you also add this information as a comment above the function (preferably in doxygen style)?

**Comment by phoebusm** [comment]
No deterioration of performance observed after adding the lock:
Average time taken for `read`
Before adding the lock: 0.027s
After adding the lock: 0.023s

File: python/tests/integration/arcticdb/test_arctic_batch.py
Comment (poodlewars) [inline python/tests/integration/arcticdb/test_arctic_batch.py:1345]: Increase these thread counts?

What is the point of the queue (which isn't synchronised correctly), why not just let the exception bubble up and fail the test?

File: cpp/arcticdb/version/version_store_api.hpp
Comment (poodlewars) [inline cpp/arcticdb/version/version_store_api.hpp:348]: I'm confused, I can't see where this is defined, or used?

File: cpp/arcticdb/storage/python_bindings.cpp
Comment (poodlewars) [inline cpp/arcticdb/storage/python_bindings.cpp:152]: Why does this method need the mutex, and not the other methods in this class?

File: python/tests/integration/arcticdb/test_arctic_batch.py
Comment (phoebusm) [reply python/tests/integration/arcticdb/test_arctic_batch.py:1345]: I can increase the thread count. Yet according to my test, current setting is sufficient to repeating the issue. I would like to keep the additional time needed for running the test as minimum as possible.

File: cpp/arcticdb/version/version_store_api.hpp
Comment (phoebusm) [reply cpp/arcticdb/version/version_store_api.hpp:348]: It is used as `py::call_guard<SingleThreadMutexHolder>()` while defining `pybind` function. 
```
m.def("foo", foo, py::call_guard<T>());
```
is equal to 
```
m.def("foo", [](args...) {
    T scope_guard;
    return foo(args...); // forwarded arguments
});
```

File: cpp/arcticdb/async/python_bindings.cpp
Comment (phoebusm) [reply cpp/arcticdb/async/python_bindings.cpp:11]: Oops sorry will be remove

File: cpp/arcticdb/version/python_bindings.cpp
Comment (phoebusm) [reply cpp/arcticdb/version/python_bindings.cpp:410]: I have thought about that but the `call_guard` is only needed for functions involves folly::future, which are not the majority of the case. So at the end of the day, developers still need to differentiate whether they need the guard or not. Then, in your case, assigning the new subclass of `py::class_`. To me, it is not a big different to adding the `py::call_guard`.
Or, if you agree adding the mutex to add functions, it will justify this idea, I think.

File: cpp/arcticdb/storage/python_bindings.cpp
Comment (phoebusm) [reply cpp/arcticdb/storage/python_bindings.cpp:152]: I think I should have put the explanation in the description of this PR.
So basically, my understanding is, this will be short-lived patch. So my idea is keeping it safe for the short term:
1. No major rehaul
2. Add mutex to all functions in `PythonVersionStore`
3. If functions use folly::Future, which, in my eyes, can be converted async task easily. So 
`remove_key` is called by `remove_library_config`, which uses `folly::Future`, mutex is necessary

File: python/tests/integration/arcticdb/test_arctic_batch.py
Comment (poodlewars) [reply python/tests/integration/arcticdb/test_arctic_batch.py:1345]: Sounds good, the minimal amount to repro the issue is fine.

File: cpp/arcticdb/version/python_bindings.cpp
Comment (poodlewars) [reply cpp/arcticdb/version/python_bindings.cpp:410]: I was thinking just for `PythonVersionStore`, but I can see that my idea might add about as much complexity as it removes. Let's leave it as-is for now.

File: cpp/arcticdb/version/version_store_api.hpp
Comment (phoebusm) [reply cpp/arcticdb/version/version_store_api.hpp:348]: Oops I see your point now. Removing

**Review by poodlewars — APPROVED** [review]

**Review by poodlewars — CHANGES_REQUESTED** [review]
Worried about forking? Then new process will get the mutex state from the static one in `pybind_mutex.hpp`, which may be locked. How will it ever get unlocked? Do we need some mechanism to have a per-process mutex? Does pybind have anything to help us here?

**Review by poodlewars — APPROVED** [review]

**Comment by phoebusm** [comment]
> Worried about forking? Then new process will get the mutex state from the static one in `pybind_mutex.hpp`, which may be locked. How will it ever get unlocked? Do we need some mechanism to have a per-process mutex? Does pybind have anything to help us here?

Handling for multiprocess added

**Pull request merged**
